File: cpp/p7-reverse-integer.cpp
#include <cstdint>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int reverse(int x)
    {
        int res = 0;
        while (x != 0) {
            if (x < 0 && (INT32_MIN - x % 10) / 10 > res) {
                return 0;
            }
            if (x > 0 && (INT32_MAX - x % 10) / 10 < res) {
                return 0;
            }
            res = res * 10 + x % 10;
            x /= 10;
        }
        return res;
    }
};

// First cut at solution.
class Solution2 {
public:
    int reverse(int x)
    {
        int64_t ret = 0;

        while (true) {
            ret += x % 10;
            x /= 10;
            if (x == 0) {
                break;
            }
            ret *= 10;
        }
        if (ret >= 0) {
            return (ret <= INT32_MAX) ? ret : 0;
        }
        return (ret >= INT32_MIN) ? ret : 0;
    }
};

int main(int argc, char* argv[])
{
    struct TestCase {
        int num;
        int exp;
    };
    const TestCase testCases[] = {
        { 0, 0 },
        { 123, 321 },
        { 1463847412, 2147483641 },
        { 1563847412, 0 },
        { -1463847412, -2147483641 },
        { -1563847412, 0 },
        { -123, -321 },
        { 120, 21 },
    };

    Solution s;
    for (const auto& tc : testCases) {
        const int ans = s.reverse(tc.num);
        if (tc.exp != ans) {
            cout << "FAIL. num: " << tc.num
                 << ", ans: " << ans
                 << ", exp: " << tc.exp << "\n";
        }
    }
    return 0;
}

File: py/p7-reverse-integer.py
class Solution:
    def reverse(self, x: int) -> int:
        ret = 0
        isnegative = x < 0
        x = abs(x)
        while True:
            ret += x % 10
            x = x // 10
            if x == 0:
                break
            ret *= 10

        # Return 0 if result is an int32 underflow or overflow.
        if isnegative:
            ret *= -1
            if ret < -2147483648:
                return 0
        elif ret > 2147483647:
            return 0

        return ret


if __name__ == '__main__':

    s = Solution()
    tests = (
        (123, 321),
        (7463847412, 2147483647),
        (8463847412, 0),
        (-8463847412, -2147483648),
        (-9463847412, 0),
        (-123, -321),
        (120, 21),
    )

    for test in tests:

        num = test[0]
        exp = test[1]
        ans = s.reverse(num)
        if exp != ans:
            print(f'FAIL. num: {test[0]}, exp: {exp}, ans: {ans}')
